proxy/src/transport/tls/config.rs

```diff
@@ -0,0 +1,142 @@
+use std::{
+    fs::File,
+    io::{self, Cursor, Read},
+    path::PathBuf,
+    sync::Arc,
+};
+
+use super::{
+    cert_resolver::CertResolver,
+
+    rustls,
+    untrusted,
+    webpki,
+};
+
+/// Not-yet-validated settings that are used for both TLS clients and TLS
+/// servers.
+///
+/// The trust anchors are stored in PEM format because, in Kubernetes, they are
+/// stored in a ConfigMap, and until very recently Kubernetes cannot store
+/// binary data in ConfigMaps. Also, PEM is the most interoperable way to
+/// distribute trust anchors, especially if it is desired to support multiple
+/// trust anchors at once.
+///
+/// The end-entity certificate and private key are in DER format because they
+/// are stored in the secret store where space utilization is a concern, and
+/// because PEM doesn't offer any advantages.
+#[derive(Debug)]
+pub struct CommonSettings {
+    /// The trust anchors as concatenated PEM-encoded X.509 certificates.
+    pub trust_anchors: PathBuf,
+
+    /// The end-entity certificate as a DER-encoded X.509 certificate.
+    pub end_entity_cert: PathBuf,
+
+    /// The private key in DER-encoded PKCS#8 form.
+    pub private_key: PathBuf,
+}
+
+/// Validated configuration common between TLS clients and TLS servers.
+pub struct CommonConfig {
+    cert_resolver: Arc<CertResolver>,
+}
+
+/// Validated configuration for TLS servers.
+#[derive(Clone)]
+pub struct ServerConfig(pub(super) Arc<rustls::ServerConfig>);
+
+#[derive(Debug)]
+pub enum Error {
+    Io(PathBuf, io::Error),
+    FailedToParsePrivateKey,
+    FailedToParseTrustAnchors(Option<webpki::Error>),
+    EmptyEndEntityCert,
+    EndEntityCertIsNotValid(webpki::Error),
+    InvalidPrivateKey,
+    TimeConversionFailed,
+}
+
+impl CommonConfig {
+    /// Loads a configuration from the given files and validates it. If an
+    /// error is returned then the caller should try again after the files are
+    /// updated.
+    ///
+    /// In a valid configuration, all the files need to be in sync with each
+    /// other. For example, the private key file must contain the private
+    /// key for the end-entity certificate, and the end-entity certificate
+    /// must be issued by the CA represented by a certificate in the
+    /// trust anchors file. Since filesystem operations are not atomic, we
+    /// need to check for this consistency.
+    pub fn load_from_disk(settings: &CommonSettings) -> Result<Self, Error> {
+        let trust_anchor_certs = load_file_contents(&settings.trust_anchors)
```

Comment:
This ugliness is needed because the `webpki` module wasn't optimized for dynamic configuration of CAs. We'll change the webpki API later to make this nicer.